Fix google_publish

adborden opened the pull request:
Add explicit start/end datetime objects.

Fixes https://github.com/openoakland/dos-census-events/issues/54

adborden commented:
Oh, I wanted to add some tests before we merge...

adborden commented:
Okay, this is ready for review.

mikeubell commented on census/templates/event.html:
```diff
@@ -70,6 +58,19 @@
+          <span class="usa-hint line-height-body-1 font-body-2xs">{{form.end_datetime.help_text}}</span>
+          <input type="time" name="end_datetime" required="" id="{{form.end_datetime.id_for_label}}">
+
+          <label class="usa-label" for="{{ form.recurrences.id_for_label }}">Recurrences:</label>
```
I'd like to see the recurrence stuff off by default.  Can we have it displayed only with DEBUG or something?  We are getting close to sharing this with our partners and others (I have to be part a webinar for the Census Lab on Friday).  Since I cannot state what the rules are for the interaction of the start_datetime/end_datetime and recurrences, I would rather not have to try to explain it.

adborden replied:
FWIW, I'm pretty sure the recurrences field is working. I'll put it behind a feature flag. You can enable it by appending a query parameter `/submit/?enable_recurrence=1`

adborden commented:
Okay this is ready for another review. I encourage folks to try it with the feature flag. I think that just by switching the order (putting start/end fields _before_ recurrence) primes your brain and it's much more intuitive than before.

mikeubell commented:
Not sure what you mean by changing the order makes in more intuitive.  I do not know what it means to set a start/end date and a recurrence.  For example, if you set a start/end date to next tuesday and say the event recurs every Thursday what dates does the event get scheduled for?  Does it start this coming Thursday or is it the Thursday after the specified Tuesday?  Or is this an error since you said it is on a Tuesday and recurs on Thursday?  If you can document the functionality I would feel better about releasing the feature.

mikeubell commented:
This change breaks the generation of the /pending list.  I think this is because it currently looks at the recurances field to get the start date. (You should rebase to get the fix of to the approval_status default)

adborden commented:
Can you write a test for the /pending list?

n8snyder commented:
> Can you write a test for the /pending list?

I can do this.

mikeubell commented:
Thanks!

On Wed, Nov 13, 2019, 9:39 PM n8snyder <notifications@github.com> wrote:

> Can you write a test for the /pending list?
>
> I can do this.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/openoakland/dos-census-events/pull/57?email_source=notifications&email_token=ABTSH6HGC57VMPVBRJHYXO3QTTQAXA5CNFSM4JLYFEB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEAU2LY#issuecomment-553733423>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTSH6DVBFXCT4GANGV5TV3QTTQAXANCNFSM4JLYFEBQ>
> .
>

mikeubell commented on census/models.py:
```diff
@@ -23,9 +23,9 @@ def first_date(self):
-    start_time = models.TimeField(help_text="What time does this event start? HH:MM:AM/PM")
-    end_time = models.TimeField(help_text="What time does the event end? HH:MM:AM/PM")
+    recurrences = RecurrenceField(default=None, blank=True, help_text="This event occurs more than once.")
+    start_datetime = models.DateTimeField(help_text="When does the event start?")
```
We don't seem to get a calendar widget on the datetime fields.  I thought we had this before.

adborden replied:
Yes, when we were using Django's built in widgets, but now we are writing custom HTML ourselves for uswds and we'll have to do some work to get the Widget working. I won't be able to do that as part of this PR.

adborden commented:
> For example, if you set a start/end date to next tuesday and say the event recurs every Thursday what dates does the event get scheduled for?

This would be good to test in a usability session. What would you expect to happen? 

I would expect exactly what you said. It would start on Tuesday and repeat weekly on Thursdays.

Spoiler, you can test this in Google Calendar with custom recurrence. You can also start on Monday and set a weekly recurrence for Thursdays and Fridays. It's all valid in the icalendar spec.

adborden commented:
I think we can steal Google's idea and hide the recurrence widget behind a drop down. The most common scenarios would be in the drop-down or you could select custom to get the Widget for complex recurrence rules.

mikeubell commented on census/tests/test_google_publish.py:
```diff
@@ -0,0 +1,121 @@
+
+        call_command('google_publish')
+
+        mock_google_publish_event.assert_not_called()
```
I don't see where in google_publish.py we only publish APPROVED events.  The code is 
 for event in Event.objects.all(): 
And there is no check for EventApprovalStatus there or in google_publish_event.  Why does this test work then?

mikeubell commented:
> I think we can steal Google's idea and hide the recurrence widget behind a drop down. The most common scenarios would be in the drop-down or you could select custom to get the Widget for complex recurrence rules.

This would be good.  I have done some testing and I see that the startdate limits when the recurrence starts, **except** that you can specify specific dates in the recurrence that is before the startdate.  The enddate does **not** affect the recurrence.  It would be good if we can force the widget to default to requiring an end date or a count on the recurrence to avoid infinite events.

mikeubell commented:
> Can you write a test for the /pending list?

I think you need to change event.first_date() to something like:
return self.start_datetime.date()

adborden commented:
> It would be good if we can force the widget to default to requiring an end date or a count on the recurrence to avoid infinite events.

What's the issue with infinite events?

mikeubell commented:
If you don't specify an end date or number of recurring events Google will
schedule them out for ever.

On Fri, Nov 15, 2019, 12:59 PM Aaron D Borden <notifications@github.com>
wrote:

> It would be good if we can force the widget to default to requiring an end
> date or a count on the recurrence to avoid infinite events.
>
> What's the issue with infinite events?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/openoakland/dos-census-events/pull/57?email_source=notifications&email_token=ABTSH6HVBZD2NYUEREBF6UTQT4ESPA5CNFSM4JLYFEB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEGWKFQ#issuecomment-554525974>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTSH6FHB64KWSTG4UIFKYTQT4ESPANCNFSM4JLYFEBQ>
> .
>

adborden commented:
So, if you had an event like OpenOakland Hack Night, that recurs every Tuesday indefinitely, I would be forced to put a limit on the recurrence?

mikeubell commented:
> So, if you had an event like OpenOakland Hack Night, that recurs every Tuesday indefinitely, I would be forced to put a limit on the recurrence?

Well the census ends sometime in 2020, its not likely they will have events after that.  Its not a big issue.

mikeubell commented:
> > Can you write a test for the /pending list?
> 
> I think you need to change event.first_date() to something like:
> return self.start_datetime.date()

Actually something like the below is needed, otherwise the dates are reported relative to UTC.
      def first_date(self):                                                          
          d = self.start_datetime.astimezone(timezone(settings.TIME_ZONE))           
          return d.date()

adborden commented:
We might want to move the conversation on recurrence to a GH issue.

I just rebased on master, but all the tests are still passing, even for the new pending test case...

adborden commented:
I updated the pending list view to test both the recurring case and the non-recurring case, which would've caught the first_date issue. I removed `first_date` since it is equivalent to start_datetime.

adborden commented:
I think there might still be a bug here with the form not including the timezone info. Since in general, the calendar will always operate in local time, it probably will never come up. I could use some Django help to debug. This manifests as a warning when you run the test:

```
/home/adborden/.local/share/virtualenvs/dos-census-events-JxEkV-SB/lib/python3.6/site-packages/django/db/models/fields/__init__.py:1427: RuntimeWarning: DateTimeField Event.end_datetime received a naive datetime (2019-11-01 17:30:00) while time zone support is active.
  RuntimeWarning)
/home/adborden/.local/share/virtualenvs/dos-census-events-JxEkV-SB/lib/python3.6/site-packages/django/db/models/fields/__init__.py:1427: RuntimeWarning: DateTimeField Event.start_datetime received a naive datetime (2019-11-01 16:00:00) while time zone support is active.
  RuntimeWarning)
```

adborden commented:
Oops, I forgot to look into that.
